### PR TITLE
TYP: Update type hints for ExtensionArray and ExtensionDtype

### PR DESCRIPTION
**Pull request by Dr-Irv**
Modifications to reduce error reports from `pyright --verifytypes` related to `ExtensionArray` and `ExtensionDtype`.

Some notes:
1. Had to use `Any` in a number of places, because an `ExtensionArray` can contain any type
2. Because `numpy` doesn't have a released version with types, `pyright` complains whenever we work with `numpy` types, so for type checking purposes, I redefine `np.dtype` and `np.ndarray` to a dummy type.  When `numpy` 1.20 is released, we can most likely remove that.  Alternatively, I can find out if there is a way to see if `pyright` has something like `TYPE_CHECKING` to bracket those changes.

I anticipate these changes might be controversial, so constructive criticism is welcome!

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
```diff
-        method: Optional[Literal["backfill", "bfill", "ffill", "pad"]] = None,
+        method: Optional[Literal[backfill, bfill, ffill, pad]] = None,
```
Looks like a bug in `no-string-hints`, sorry about that, will test + update!

**Comment by simonjayhawkins** [comment]
> 2\. When `numpy` 1.20 is released

@Dr-Irv can you update

**Comment by Dr-Irv** [comment]
> > 2. When `numpy` 1.20 is released
> 
> @Dr-Irv can you update

Ha!  It was just released in pypl on 1/30/21, which is when I pushed this!  Waiting for it to be on conda and then will update.

**Comment by Dr-Irv** [comment]
> > 2. When `numpy` 1.20 is released
> 
> @Dr-Irv can you update

I did try this, and now `mypy pandas` reports more errors because `numpy` 1.20 has typing.  #39513 

I'm asking the `pyright` people to see if there is a different workaround.  Ideally, I'd like to tell pyright "process pandas, but ignore numpy"

**Comment by Dr-Irv** [comment]
> I'm asking the `pyright` people to see if there is a different workaround. Ideally, I'd like to tell pyright "process pandas, but ignore numpy"

You can now use `pyright` to find pandas symbols that are missing types by using `pyright --verifytypes pandas!` (The `!` means don't look at imported libraries).  Based on that (and including this PR), we have 75.7% completeness for pandas.

**Comment by Dr-Irv** [comment]
> Looks like a bug in `no-string-hints`, sorry about that, will test + update!
> 
> Have updated and pushed a new tag, the version in pandas' pre-commit-config will get updated tomorrow morning as part of the scheduled weekly update

@MarcoGorelli Has this been done yet?  I'm still getting the pre-commit failure

**Comment by MarcoGorelli** [comment]
> > Looks like a bug in `no-string-hints`, sorry about that, will test + update!
> > Have updated and pushed a new tag, the version in pandas' pre-commit-config will get updated tomorrow morning as part of the scheduled weekly update
> 
> @MarcoGorelli Has this been done yet? I'm still getting the pre-commit failure

yup - if you approve/merge #39524 and then merge, then it'll pass

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -49,6 +51,9 @@
 from pandas.core.missing import get_fill_func
 from pandas.core.sorting import nargminmax, nargsort
 
+if TYPE_CHECKING:
+    from typing import Literal
```
Comment (MarcoGorelli) [inline pandas/core/arrays/base.py:55]: Why only import this one if TYPE_CHECKING? (asking because I don't know, not because I think it's wrong)

Comment (Dr-Irv) [reply pandas/core/arrays/base.py:55]: `Literal` was only used in one other place `core/frame.py`, so I followed the pattern there.  I think the issue is that `Literal` was introduced in python 3.8, so doing it with if `TYPE_CHECKING` makes sure that it is compatible with python 3.7 .

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -667,7 +672,12 @@ def argmax(self, skipna: bool = True) -> int:
             raise NotImplementedError
         return nargminmax(self, "argmax")
 
-    def fillna(self, value=None, method=None, limit=None):
+    def fillna(
+        self,
+        value: Optional[Union[Any, ArrayLike]] = None,
+        method: Optional[Literal["backfill", "bfill", "ffill", "pad"]] = None,
```
Comment (MarcoGorelli) [inline pandas/core/arrays/base.py:678]: I've not followed everything in the typing discussions so sorry if this has been discussed already, but is the idea that these will one day be available to users (e.g. by including a `py.typed` file)?

If so, then I wonder whether this might be too restrictive - e.g. if someone has `method: str` and they tried passing it here, they'd get an error like
```
Argument 1 to "foo" has incompatible type "str"; expected "Literal['backfill', 'bfill', 'ffill', 'pad']"  [arg-type]
```

If users are expected to use these literals, then perhaps `Literal["backfill", "bfill", "ffill", "pad"]` should be an alias (which users could import)?

Comment (Dr-Irv) [reply pandas/core/arrays/base.py:678]: Yes, the idea is that these would be available to users.

But for something like `fillna`, I would assume that most people wouldn't pass a variable `method` as the argument, i.e., you're more likely to write `df.fillna(method = "ffill")` than 
```
def myfunc(df: pd.DataFrame, mymethod:str):
    df.fillna(method=mymethod)
```

Comment (MarcoGorelli) [reply pandas/core/arrays/base.py:678]: True - besides, this error message is perfectly clear so this is prob OK :+1:

File: pandas/core/arrays/base.py [hunk 3]
```diff
@@ -712,7 +722,7 @@ def fillna(self, value=None, method=None, limit=None):
             new_values = self.copy()
         return new_values
 
-    def dropna(self):
+    def dropna(self) -> ExtensionArray:
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:725]:
```suggestion
    def dropna(self: ExtensionArrayT) -> ExtensionArrayT:
```

and then the cast in SparseArray._reduce can be removed.

but will produce `pandas/core/arrays/base.py:733: error: Incompatible return value type (got "Union[ExtensionArray, Any]", expected "ExtensionArrayT")  [return-value]` since the return type of `ExtensionArray.__getitem__` is `Union[ExtensionArray, Any]` and this needs to be overloaded.

File: pandas/core/arrays/base.py [hunk 4]
```diff
@@ -776,7 +786,7 @@ def shift(self, periods: int = 1, fill_value: object = None) -> ExtensionArray:
             b = empty
         return self._concat_same_type([a, b])
 
-    def unique(self):
+    def unique(self) -> ExtensionArray:
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:789]: use ExtensionArrayT (return is `self._from_sequence...`)

File: pandas/core/arrays/base.py [hunk 5]
```diff
@@ -1002,7 +1017,9 @@ def factorize(self, na_sentinel: int = -1) -> Tuple[np.ndarray, ExtensionArray]:
 
     @Substitution(klass="ExtensionArray")
     @Appender(_extension_array_shared_docs["repeat"])
-    def repeat(self, repeats, axis=None):
+    def repeat(
+        self, repeats: Union[int, Sequence[int]], axis: Literal[None] = None
+    ) -> ExtensionArray:
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:1022]: same

File: pandas/core/arrays/base.py [hunk 6]
```diff
@@ -667,7 +672,12 @@ def argmax(self, skipna: bool = True) -> int:
             raise NotImplementedError
         return nargminmax(self, "argmax")
 
-    def fillna(self, value=None, method=None, limit=None):
+    def fillna(
+        self,
+        value: Optional[Union[Any, ArrayLike]] = None,
+        method: Optional[Literal["backfill", "bfill", "ffill", "pad"]] = None,
+        limit: Optional[int] = None,
+    ) -> ExtensionArray:
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:680]: same

File: pandas/core/arrays/base.py [hunk 7]
```diff
@@ -787,7 +797,12 @@ def unique(self):
         uniques = unique(self.astype(object))
         return self._from_sequence(uniques, dtype=self.dtype)
 
-    def searchsorted(self, value, side="left", sorter=None):
+    def searchsorted(
+        self,
+        value: ArrayLike,
+        side: Optional[Literal["left", "right"]] = "left",
+        sorter: Optional[ArrayLike] = None,
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:804]: `ArrayLike` is a typevar, so the type of `value` and `sorter` must be the same otherwise I _think_ we would get mypy errors. (although we may not get errors while np.ndarray resolves to any). 

(this is another area where api testing could be useful #40202)

File: pandas/core/arrays/base.py [hunk 8]
```diff
@@ -787,7 +797,12 @@ def unique(self):
         uniques = unique(self.astype(object))
         return self._from_sequence(uniques, dtype=self.dtype)
 
-    def searchsorted(self, value, side="left", sorter=None):
+    def searchsorted(
+        self,
+        value: ArrayLike,
+        side: Optional[Literal["left", "right"]] = "left",
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:803]: `Optional` not needed. within the method body, `side` will be a `str` and not `None`

File: pandas/core/dtypes/base.py [hunk 9]
```diff
@@ -452,4 +451,4 @@ def find(
         return None
 
 
-registry = Registry()
+registry: Registry = Registry()
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/base.py:454]: why is this needed?

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @Dr-Irv

File: pandas/core/arrays/base.py (hunk 3 above)
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:725]: I made that change and added overloads on `__getitem__()`

File: pandas/core/arrays/base.py (hunk 7 above)
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:804]: I don't see how we can make them the same when we are allowing `sorter` to be `None`.  I am passing the `mypy` tests with the code as above.

File: pandas/core/arrays/base.py (hunk 4 above)
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:789]: Actually, needs to be `def unique(self: ExtensionArrayT) -> ExtensionArrayT`

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:454]: I think that the `pylance` type checker wanted that.  I need to investigate it.

**Comment by MarcoGorelli** [comment]
Nice. OK, so it seems this is a case where we don't need the extra overload we had to put in #40197 (which might only be necessary if have overloads for specific instances of the base type. I.e. if you overload `bool` with `Literal[True]` and `Literal[False]` you also need an overload for `bool`, but if you overload `Union[bool, int]` with `bool` and `int` then you don't need an extra overload) 

e.g.
```python
from typing import overload, Union, List, Any

MyType = List[str]


class Animal:
    
    
    @overload
    def foo(self, key: int) -> Any: ...
    
    @overload
    def foo(self, key: str) -> MyType: ...
    
    def foo(self, key: Union[int, str]) -> Union[MyType, Any]:
        ...

key: Union[int, str]
reveal_type(Animal().foo(key))  # Revealed type is 'Union[Any, builtins.list[builtins.str]]'
```

https://mypy-play.net/?mypy=latest&python=3.9&gist=c9bd0a94f79360506940aaf8274356dc

File: pandas/core/arrays/base.py [hunk 10]
```diff
@@ -770,13 +795,13 @@ def shift(self, periods: int = 1, fill_value: object = None) -> ExtensionArray:
         )
         if periods > 0:
             a = empty
-            b = self[:-periods]
+            b: ExtensionArrayT = self[:-periods]
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:798]: this shouldn't be needed. `Revealed type is 'ExtensionArrayT`-1'` without the variable annotation. That's sort of the point of adding the overloads, so that the return types are known.

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @Dr-Irv for the updates

File: pandas/core/arrays/base.py (hunk 7 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:819]: sorry, didn't explain properly. The types are saying that the types are the same if `sorter` is not None. We don;t want that.

> I am passing the `mypy` tests with the code as above.

that's because np.ndarry resolves to `Any`. the sooner #39513 is resolved (i.e #36092 gets merged, if no-one is going to suggest a practical alternative) the better. (after over 6 months on that PR, I am almost done with typing for pandas, its getting tiresome (and very time consuming) keeping #36092 up-to-date just to see more hidden errors appearing daily.

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (MarcoGorelli) [reply pandas/core/dtypes/base.py:454]: Would it work to use stricter flags like
```
disallow_incomplete_defs = true
disallow_untyped_defs = true
```
to discover missing annotations, so then we don't have to add type hints which are redundant for `mypy`?

File: pandas/core/arrays/base.py (hunk 8 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:818]: > I don't see that. `searchsorted` calls `numpy.searchsorted` which allows `None` as an argument.

it's optional. None raises `TypeError`

```
>>> 
>>> np.searchsorted([1,2,3,4,5], 3)
2
>>> 
>>> np.searchsorted([1,2,3,4,5], 3, side='right')
3
>>> 
>>> np.searchsorted([1,2,3,4,5], 3, side=None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 5, in searchsorted
  File "/home/simon/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 1343, in searchsorted
    return _wrapfunc(a, 'searchsorted', v, side=side, sorter=sorter)
  File "/home/simon/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 55, in _wrapfunc
    return _wrapit(obj, method, *args, **kwds)
  File "/home/simon/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 44, in _wrapit
    result = getattr(asarray(obj), method)(*args, **kwds)
TypeError: search side must be str, not NoneType
>>> 
```

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py:454]: we may have already have issues for these. mypy has flags to activate at milestones in the typing journey and mypy also generates reports. (we don't really need to use pyright)

differences of opinion will arise when the motivations are typing public apis vs in-line typing for code consistency/robustness.

from https://blog.zulip.com/2016/10/13/static-types-in-python-oh-mypy/

"For Zulip, getting clean --check-untyped-defs output took me about 2 days of hard work" - it too us over a year.

Fully annotating a large codebase:
Phase 1: Annotate the core library - we need stubs for the cython files and to get the numpy types in play[1].
Phase 2: Annotate the bulk of the codebase. - 
Phase 3: Get to 100%. - use --disallow-untyped-defs to lockdown progress

so there is much todo to complete phase 1 before we can be confident in the typing of the public api.

[1] https://github.com/pandas-dev/pandas/pull/39501#discussion_r590339840 would be picked up by mypy, i.e. trying to pass `None` through to numpy.searchsorted.

File: pandas/core/arrays/base.py (hunk 10 above)
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:798]: `mypy` is complaining if I don't include that.  I opened up a `mypy` case and it's because we created the `ExtensionTypeT` type, which they say we don't need.  I'm going to experiment with that.

File: pandas/core/dtypes/base.py (hunk 9 above)
Comment (Dr-Irv) [reply pandas/core/dtypes/base.py:454]: Actually, I think the right solution here is to make `registry` be `_registry`, because it is not part of the public interface.  What `pyright --verifytypes` does is verify that anything that is in the public interface is properly typed.  I will make that change.

File: pandas/core/arrays/base.py (hunk 10 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:798]: > `mypy` is complaining if I don't include that

can you post the output. i'm not seeing that.

(what I am seeing is pylance saying the code is unreachable!)

Comment (Dr-Irv) [reply pandas/core/arrays/base.py:798]: Solution suggested by mypy folks seems to be to avoid `ExtensionArrayT` as much as possible.  I've changed it to only use it twice in `pandas/core/arrays/base.py` .

File: pandas/core/arrays/base.py [hunk 11]
```diff
@@ -282,8 +288,20 @@ def _from_factorized(cls, values, original):
     # Must be a Sequence
     # ------------------------------------------------------------------------
 
+    @overload
+    def __getitem__(self: ExtensionArrayT, item: int) -> Any:
+        ...
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:293]: Needed to handle things correctly in the overloads in `pandas/core/arrays/_mixins.py`

File: pandas/core/arrays/base.py [hunk 12]
```diff
@@ -282,8 +288,20 @@ def _from_factorized(cls, values, original):
...
+
+    @overload
+    def __getitem__(self: ExtensionArrayT, item: slice) -> ExtensionArrayT:
+        ...
+
+    @overload
+    def __getitem__(self: ExtensionArrayT, item: np.ndarray) -> ExtensionArrayT:
+        ...
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:301]: done

File: pandas/core/arrays/base.py [hunk 13]
```diff
@@ -282,8 +288,20 @@ def _from_factorized(cls, values, original):
...
+
     def __getitem__(
-        self, item: Union[int, slice, np.ndarray]
+        self: ExtensionArrayT, item: Union[int, slice, np.ndarray]
     ) -> Union[ExtensionArray, Any]:
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:305]: Not needed.  Using `ExtensionArray` wherever possible.

File: pandas/core/arrays/base.py [hunk 14]
```diff
@@ -712,7 +735,7 @@ def fillna(self, value=None, method=None, limit=None):
             new_values = self.copy()
         return new_values
 
-    def dropna(self):
+    def dropna(self) -> ExtensionArrayT:
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:738]: Seems to be OK without it.

File: pandas/core/arrays/base.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:293]: was able to remove it

File: pandas/core/arrays/base.py [hunk 15]
```diff
@@ -667,7 +681,12 @@ def argmax(self, skipna: bool = True) -> int:
             raise NotImplementedError
         return nargminmax(self, "argmax")
 
-    def fillna(self, value=None, method=None, limit=None):
+    def fillna(
+        self: ExtensionArray,
```
Comment (simonjayhawkins) [inline pandas/core/arrays/base.py:685]: FYI: no need to update at the moment. if #36092 goes in first, you'll probably need to update this PR anyway.

`self` does not need a type annotation except when using a typevar.

File: pandas/core/arrays/base.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:734]: yep. tis strange that mypy is not complaining. 

from https://github.com/python/mypy/issues/10190#issuecomment-795542485

> Mypy should complain about a plain type variable as a return type if it's not used anywhere else.

File: pandas/core/arrays/base.py (hunk 15 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:685]: see https://mypy.readthedocs.io/en/stable/generics.html?highlight=self#generic-methods-and-generic-self and mypy 0.740 release notes http://mypy-lang.blogspot.com/search?updated-max=2019-11-29T11:39:00-08:00&max-results=7

> The self argument to methods without annotations is now correctly given the object’s type (instead of Any) when using check_untyped_defs (PR 7530).

Comment (Dr-Irv) [reply pandas/core/arrays/base.py:685]: I just tried removing them all from `pandas/core/arrays/base.py` and that seems to work.

File: pandas/core/arrays/base.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas/core/arrays/base.py:734]: If I switch it to `def dropna(self) -> ExtensionArray:`, then we have the problem with the sparse array call to `dropna()`, so I'm leaving it in there for now.

**Comment by Dr-Irv** [comment]
@simonjayhawkins Ugh the merge of #36092 raised a whole bunch of issues with this PR that I now have to address.

**Comment by Dr-Irv** [comment]
> @simonjayhawkins Ugh the merge of #36092 raised a whole bunch of issues with this PR that I now have to address.

@simonjayhawkins So in trying to fix these issues, if I change the declaration of `AnyArrayLike` and `ArrayLike` to be a `Union` instead of a `TypeVar`, I can get rid of a LOT of the `# type: ignore` stuff you added.  Before I start doing that, do you have an opinion on them being a `TypeVar` versus a `Union`

**Comment by simonjayhawkins** [comment]
> do you have an opinion on them being a `TypeVar` versus a `Union`

done in #40379

**Comment by Dr-Irv** [comment]
> > do you have an opinion on them being a `TypeVar` versus a `Union`
> 
> done in #40379

Oh, so maybe I will wait for you to get that merged before I go further on this PR.

**Comment by Dr-Irv** [comment]
@simonjayhawkins Updated the PR based on your latest stuff with typing, so now `pandas/core/arrays/base.py` is fully typed and that allowed me to remove a bunch of `# type; ignore` things.   Main changes from last review is an overload for `astype` and fixing `isin`.  Looking forward to your review.

**Comment by simonjayhawkins** [comment]
Thanks @Dr-Irv for working on this.

I would imagine that this PR will be easier to check if we limit the changes to `pandas/core/arrays/base.py` and the other changes are done in a separate PR(s).

There is a conflict here and I _think_ some other overlaps with other PRs removing the ignores.

some of the changes to remove ignores need more discussion than others. If ignores are being removed that's great, if ignores or changed or others added we need to review in more depth.

File: pandas/core/algorithms.py [hunk 16]
```diff
@@ -216,7 +218,7 @@ def _ensure_data(values: ArrayLike) -> Tuple[np.ndarray, DtypeObj]:
 
 
 def _reconstruct_data(
-    values: ArrayLike, dtype: DtypeObj, original: AnyArrayLike
+    values: ArrayLike, dtype: Dtype, original: AnyArrayLike
```
Comment (simonjayhawkins) [inline pandas/core/algorithms.py:221]: I think for internal low level functions we wanted to keep the dtype parameter restricted to dtype objects. The errors in this section were fixed in #40412.

I'll push a commit merging master.

File: pandas/core/arrays/datetimelike.py [hunk 17]
```diff
@@ -459,7 +460,9 @@ def view(self: DatetimeLikeArrayT) -> DatetimeLikeArrayT:
         ...
 
     @overload
-    def view(self, dtype: Optional[Dtype] = ...) -> ArrayLike:
+    def view(
+        self: DatetimeLikeArrayT, dtype: Optional[Dtype] = ...
+    ) -> DatetimeLikeArrayT:
```
Comment (simonjayhawkins) [inline pandas/core/arrays/datetimelike.py:465]: hmm, what if dtype is say 'i8'. the return type won't be the same as self

**Comment by simonjayhawkins** [comment]
> I'll push a commit merging master.

some new mypy errors and now also another conflict from a subsequent merge to master related to https://github.com/pandas-dev/pandas/pull/39501#discussion_r593767378

**Comment by Dr-Irv** [comment]
> I would imagine that this PR will be easier to check if we limit the changes to `pandas/core/arrays/base.py` and the other changes are done in a separate PR(s).

I'm not sure that it can be done alone, because `pandas/core/arrays/base.py` has the base `ExtensionArray` class, so you have to fix things upstream when fixing things there.  
> 
> There is a conflict here and I _think_ some other overlaps with other PRs removing the ignores.
> 
> some of the changes to remove ignores need more discussion than others. If ignores are being removed that's great, if ignores or changed or others added we need to review in more depth.

I removed ignores only based on when mypy said they were not needed any more.  I don't think I added any ignores.

Let me get the current version in a state that goes green.

**Comment by simonjayhawkins** [comment]
> I'm not sure that it can be done alone, because `pandas/core/arrays/base.py` has the base `ExtensionArray` class, so you have to fix things upstream when fixing things there.
> 
> >

I think that while numpy types resolved to `Any` this PR adding types to  `pandas/core/arrays/base.py` was moreless ready.

all the other changes are related to numpy types now having some meaning. I'm sure these could be done separately (and in smaller chunks)

changes to 36 files takes some review, especially as changes are made as other PRs are merged.

typing PRs need to be atomic to be easily reviewed. 

> Main changes from last review is an overload for astype and fixing isin. Looking forward to your review.

so maybe a PR for the astype fixes and a PR for the isin fixes?

**Comment by Dr-Irv** [comment]
> all the other changes are related to numpy types now having some meaning. I'm sure these could be done separately (and in smaller chunks)

If I limit things to `pandas/core/arrays/base.py`, I'm pretty sure I can't get it to be green.  Is that OK?

**Comment by Dr-Irv** [comment]
@simonjayhawkins Latest commit https://github.com/pandas-dev/pandas/pull/39501/commits/5bb24d4aeeb44f5877a0b35d8f2389b6bf140b54 still has a variety of changes.  It just takes what you merged in and should make it mypy happy.

File: pandas/core/arrays/datetimelike.py (hunk 17 above)
Comment (Dr-Irv) [reply pandas/core/arrays/datetimelike.py]: I saw you addressed that in a recent push to master

**Comment by Dr-Irv** [comment]
> so maybe a PR for the astype fixes and a PR for the isin fixes?

I'm going to create a separate PR that is effectively this one, MINUS the astype and isin.

**Comment by Dr-Irv** [comment]
Things to do once #40421 is approved:
- [ ] `registry` as private
- [ ] `astype` overloads
- [ ] `getitem` overloads
- [ ] change `TypeVar` on `DatetimeLikeScalar` to be `Union`
- [ ] arguments for `isin`

**Comment by jbrockmendel** [comment]
> Things that can be done independent of approval of #40421

Looks like you've made some progress on that checklist.  Is this PR still active?

**Comment by Dr-Irv** [comment]
> > Things that can be done independent of approval of #40421
> 
> Looks like you've made some progress on that checklist. Is this PR still active?

I will change it to draft status, as I'm using it to track the changes I would do after #40421 is approved.  Then when I finish the checklist, I'll close this one.

**Comment by simonjayhawkins** [comment]
> I will change it to draft status, as I'm using it to track the changes I would do after #40421 is approved. Then when I finish the checklist, I'll close this one.

Is this still the master PR? if so could you merge master so we can monitor progress. and more easily see the outstanding review comments.

**Comment by Dr-Irv** [comment]
> > I will change it to draft status, as I'm using it to track the changes I would do after #40421 is approved. Then when I finish the checklist, I'll close this one.
> 
> Is this still the master PR? if so could you merge master so we can monitor progress. and more easily see the outstanding review comments.

Yes.  But I'd prefer to do the merge once  #41203,  #41251, #41258 and  #41513 are merged in so we can see where we are with respect to the comments.

**Comment by simonjayhawkins** [comment]
ok. will close for now. reopen when ready.
